[SEMVER-MAJOR] Stop gzipping CLDR files

bajtos opened the pull request:
Rework the module to use plain JSON format for CLDR files and thus get rid of dependency on zlib's gunzipAsync that's not available on Node v0.10 on Windows.

Close #46

/to @Setogit please review
/cc @rmg @sam-github

rmg commented:
Well that's definitely a simpler way to pull out the gz usage than converting it all to async!

sam-github commented:
+1

@bajtos  Why is it semver-major? Isn't this still API compatible? If the API changed, can you call the change in the body of the commit message?

bajtos commented:
> Why is it semver-major? Isn't this still API compatible? If the API changed, can you call out the change in the body of the commit message?

Well, I am worried that if anybody was depending on CLDR files having `.gz` extensions and adding custom CLDR files to be picked up by strong-globalize, then they need to start using `.json` instead.

See https://github.com/strongloop/strong-globalize#language-config-customization, emphasis is mine:
> Out of box, one CLDR gz file is included in strong-globalize/cldr directory. CLDR stands for Common Locale Data Repository. The gz file contains CLDR data for the languages: de, en, es, fr, it, ja, ko, pt, ru, zh-Hans, and zh-Hant. **In the installation of strong-globalize in your package for your production deployment, you can replace the out-of-box gz file entirely, or add extra CLDR data to the cldr directory.** There are approximtely 450 locales (language/culture variations) defined in the Unicode CLDR. Among them, there are 40+ variations of French and 100+ variations of English.

> (...)

> In runtime, strong-globalize dynamically loads to memory just the CLDR data required for the specific language by setLanguage(). **First, it examines all the gz files under cldr directory in alphabetical order**, then searches for the language. If the language is defined in two or more gz files, duplicate objects will be overwritten in the examination order.


To be honest, I find the practice of adding files deep into `node_modules` subtree ( `node_modules/strong-globalize/cldr`) a bit weird and rather difficult to use, so I think there will be very few (if any) users affected by this change, but I prefer to be safe and bump up the major version.

Should I describe this change in the commit message (an PR description) then? Something along the lines "if you are using custom CLDR files, please stop gzipping them and use plain JSON instead".

rmg commented:
@bajtos @sam-github that sounds to me like a legitimate reason to bump the major version - regardless of whether anyone is actually using it or not.

bajtos commented:
I have updated the commit message and added the following paragraph:

> Breaking change: when scanning for custom CLDR files, strong-globalize we no longer recogizes `.gz` files. If you are relying on this feature, then ungzip your CLDR file and rename them to `.json`.

@sam-github @rmg @Setogit PTAL

rmg commented on README.md:
```diff
@@ -79,7 +79,7 @@ StrongLoop Globalize CLI and API
 # Architecture
 
-`strong-globalize` is built on top of two foundation layers: Unicode CLDR and jquery/globalize.  The Unicode CLDR provides key building blocks for software to support the world's languages, with the largest and most extensive standard repository of locale data available.  jquery/globalize is a JavaScript library for internationalization and localization that leverages the Unicode CLDR JSON data.  ...
+`strong-globalize` is built on top of two foundation layers: Unicode CLDR and jquery/globalize.  The Unicode CLDR provides key building blocks for software to support the world's languages, with the largest and most extensive standard repository of locale data available.  jquery/globalize is a JavaScript library for internationalization and localization that leverages the Unicode CLDR JSON data.  ...
```
I realize it was your editor that automatically trimmed the trailing whitespace.. but since it did, maybe we should just go ahead and wrap these long lines while we're at it?

Either that or remove the whitespace change from the commit, maybe?

bajtos replied:
I am going to remove whitespace changes from my patch.

rmg commented on lib/helper.js:
```diff
@@ -15,10 +15,6 @@ var mkdirp = require('mkdirp');
 var path = require('path');
 var nodeVersion = process.version.replace(
```
This appears to be no longer used.

rmg commented:
@bajtos this appears to remove all uses of `nodeVersion` in the various files. Could you remove it entirely?

bajtos commented:
Apparently @Setogit has already removed the dependency on `node-zlib-backport` in #49, thus this patch is no longer relevant.

chandadharap commented:
@Setogit The dependency was removed after the patch was submitted. better practice would have been to accept the patch.

rmg commented:
I took a second look at how the CLDR data is loaded and I no longer think this is actually a breaking change. The workflow prescribed by the strong-globalize docs for using custom CLDR data requires you to modify strong-globalize itself anyway. If they wish to update to a newer version of strong-globalize they need to make all of their changes again anyway. This workflow lends itself more to forking this repo anyway instead of installing from npm, which takes it completely outside the scope of semantic versioning.

Setogit commented:
@bajtos: we discussed further and agreed that moving compressed CLDR to uncompressed CLDR is a right thing, i.e., your this PR.  A specific work flow I did not consider when i implemented #49 and why #49 is incomplete, as @rmg pointed out, is the following:

`npm install` dependent packages including strong-globalize, then 'npm publish'.

For example, if the build operation is done on Node v0.10, #49 solution (postinstall script removing either json or gz CLDR depending on the Node version the build operation is done), then the strong-globalize package right before `npm publish` does not contain 'gz' file.  With missing 'gz'  the published packaged works only on Node v0.10.  If you run the build operation on Node v4, the resulted package won't work for Node v0.10.

The remaining change i have on my plate is to expand the language set of default CLDR JSON to meet most globalization needs to minimize the possibility of client's having to fork.
